src/Text/Gigaparsec/Internal.hs

```diff
@@ -92,10 +108,20 @@ instance Monad Parsec where
 
 instance Alternative Parsec where
   empty :: Parsec a
-  empty = undefined -- TODO:
+  empty = Parsec $ \st _ err -> err st
 
   (<|>) :: Parsec a -> Parsec a -> Parsec a
-  (<|>) = undefined -- TODO:
+  Parsec p <|> Parsec q = Parsec $ \st ok err ->
+    let ok' x st'
+          | consumed st' = ok x st'
+          | otherwise    = ok x st
+          --  ^ revert to old st.consumed if p didn't consume
+        err' st'
+          | consumed st' = err st'
+          --  ^ fail if p failed *and* consumed
+          | otherwise    = q st ok err
```

Comment:
```suggestion
    | otherwise = q st' ok err
```

Reason being, this allows `st` to be garbage collected sooner, avoiding potential space leak.

In a similar vein I'd probably hoist `where !initConsumed = consumed st` or something so that `st` can be entirely dropped.

Reply:
Addressed in 1fd30ee